### ISSUE TITLE
Using linear regression multiple times on the same data gives me very different predictions

### ISSUE THREAD

**Issue by shawnbrar** [mlpack/mlpack#2567]
I have recently started to use `mlpack`library, so I am still learning it. The problem is when I am using the Linear Regression method on the same dataset, each time it give me back very different predictions. So, following the following is the code that I using in C++:-

__________________________________________________________________________________________
#include <mlpack/core.hpp>
#include <mlpack/core/data/load.hpp>
#include <mlpack/methods/linear_regression/linear_regression.hpp>
#include < vector >

int main(int argc, char* argv[]){
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;arma::mat rel_data;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;mlpack::data::Load(argv[1], rel_data, false);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;rel_data.shed_row(0); rel_data.shed_col(0);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;arma::rowvec row  = rel_data.row(1);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//remove all the columns which are 0
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        std::vector< arma::uword > idxs;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        for(arma::uword i = 0; i <= (rel_data.n_cols - 1); i++){
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;                if(row[i] == 0){
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;                        idxs.push_back(i);
                &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        arma::uvec indices(idxs);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        rel_data.shed_cols(indices);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        row = rel_data.row(3);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        rel_data.shed_row(3); rel_data.shed_row(4);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        arma::rowvec predictions;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        mlpack::regression::LinearRegression regressor;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        regressor.Train(rel_data, row);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        regressor.Predict(rel_data, predictions);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        predictions(arma::span(1, 50)).print("Prediction:");
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        return 0;
}
__________________________________________________________________________________________

This is my [dataset](https://github.com/mlpack/mlpack/files/5067054/RELIANCE.zip) which I am using. The following is my compilation command:-

`g++ -fopenmp ./c.cpp -L'pkg-config --cflags --libs mlpack armadillo'`

This compiles the code and makes an executable `a.out`. I run the program with one argument with the following command:-

`./a.out ./RELIANCE.NS.csv`

When I keep running the above command again and again it gives me very different predictions. I am posting a screenshot of my terminal below:-

![asd](https://user-images.githubusercontent.com/59639827/90093830-0f2ac100-dd4a-11ea-8dba-8ea713ac4909.png)

I would understand if the predictions were a little different each time. But these are too different. Please do tell me if this due to a problem in my code or is this due to a bug in the library.

**Comment by zoq** [comment]
Hm, I tested the code above and at least on my system I can see the same result in each run: can you save the preprocessed file e.g. using `data::Save("data.csv", rel_data);` and make sure the file is the same for each run?

**Comment by zoq** [comment]
Btw. what mlpack version are you using?

**Comment by shawnbrar** [comment]
> Hm, I tested the code above and at least on my system I can see the same result in each run: can you save the preprocessed file e.g. using `data::Save("data.csv", rel_data);` and make sure the file is the same for each run?

Yes the data seems to be the same on every run. It is the parameters of the regression model which keep changing. 

I am using the latest version of the library (3.2.2)

**Comment by shawnbrar** [comment]
I figured out the problem. Earlier I had compiled mlpack with openmp support for parallelization. Now, I've compiled it again but without openmp support. And now it is working fine. It is giving me the same result every time I run my program.

**Comment by shawnbrar** [comment]
The problem was actually with the openmp which ships with Ubunutu by default. I uninstalled it and built it from source. And now even with OpenMP, the mlpack is working correctly. Thank you.

**Comment by zoq** [comment]
Interesting, what Ubuntu version are you using?

**Comment by shawnbrar** [comment]
I am using Ubuntu 18.04.

**Comment by zoq** [comment]
Alright thanks for the info.
